man/make-coordinates-file.1
.TH MAKE-COORDINATES-FILE 1 "15 January 2018" "DEVA Project"
.SH NAME
make-coordinates-file \- convert Radiance file to a JPEG file
.SH SYNOPSIS
\fBmake-coordinates-file\fR \fIunits\fR
\fIradiance-file.hdr\fR \fIcoordinates-file\fR
.SH DESCRIPTION
Create a coordinates file for use by \fBdeva-visibility\fR. \fIunits\fR
is one of "meters", "centimeters", "feet", or "inches".
\fIradiance-file.hdr\fR is a Radiance image file with a VIEW record
corresponding to the geometry files used by \fBdeva-visibility\fR.
\fIcoordinates-file\fR is the coordinates file needed by
\fBdeva-visibility\fR.
.\" SH EXAMPLES
.SH AUTHOR
William B. Thompson


CHANGES
version 3.3.00

Most fatal errors now print out the line number and file where the error
occurred.

Added option in deva-visibility to check for false positive luminance
boundary detections.

Added feature in deva-visibility (deva-commandline) and
deva-compare-boundaries to check for very dark regions in original
image.  This is motivated by a lack of knowledge about the effects of
illumination levels below about 1 cd/m^2 on low vision acuity and
contrast sensitivity.

FFT-based Gaussian blur routine added (deva-gblur-fft). This is used by
the code that checks for very dark regions.

Added routines in deva-margin for DEVA_float image object, providing
support needed by code that finds very dark regions in original image.

Added code in visualize-hazards for displaying very dark regions and
false positive errors in hazard visualization.

Added code supporting region-of-interest (ROI) processing in
determination of Visibility Hazard Scores in deva-visibility.

Changed contrast sensitivity of presets based on personal communication
(G.E. Legge & Y-Z Xiong, Feb. 19, 2019).

Added a "legalblind" preset to the other low vision categories in
deva-filter and deva-visibility.  This specifies the U.S.
"legal blindness" acuity level along with appropriate contrast
sensitivity and color sensitivity values.  Note that this specifies a
level of visual impairment right at the legal blindness cutoff point.
It is not some sort of test of legal blindness.

Added a "normal" preset to the other low vision categories in
deva-filter and deva-visibility.  This is treated as a silent option,
not included in the prompt or man page, because the filter has not been
tested for this condition.

Cleaned up where new image objects were allocated in deva-visibility.c
and related routines.

Minor comment cleanup.

Cleanups and additions to man pages.

Various changes to CMakeLists.txt

Ran this file through a spell checker.

version 3.2.04

luminance-boundaries now issues a warning to avoid use with input images
generated by deva-filter.  The problem may be due to quantization noise
in the low precision Radiance format input image resulting in a large
number of micro edges that in turn confuse the auto thresholding
routine.

version 3.2.03

Really fix deva-visibility man page.

version 3.2.02

Minor fix to deva-visibility man page.

version 3.2.01

Added build process for MacOS and standardized directory layout for
Windows and MacOS builds.

Removed extraneous debugging command from deva-margin.c

version 3.2.00

Added two new programs: luminance-boundaries, geometry-boundaries.

In deva-filter.c (deva-filter, deva-visibility): Changed smoothing
algorithm to better deal with artifacts at the boundaries of the
smoothing windows.  Added optional debugging code to output individual
contrast bands.

In deva-margin.c:  Added compile-time option to base feathering on
average luminance over full image rather than just over the boundary
pixels.  Added compile-time option to base margin pixels on reflection
of image around the boundary, not just a fill of boundary pixels.

In dilate.c, added dt_euclid_sq_2 which used a pre-allocated output
image.

In deva-commandline.c (deva-filter, deva-visibility): Cleaned up command
summaries.  Allowed specifying smoothing_type with presets.

In ChungLeggeCSF.h: Cleanup of unused ChungLeggeCSF_PEAK_FREQUENCY
defines.

Man pages for new programs, minor updates to existing man pages.

version 3.1.02

In deva-commandline.c (deva-visibility): Added measurement type options
to usage message.

version 3.1.01

deva-compare-boundaries.c: New program to create a visualization image
comparing a standard set of boundary elements (e.g., geometry) to a
different set of boundary elements (e.g., luminance edges).  Annotation
of output image, which requires Cairo library, is includable with a
compile-time switch.

deva-add-text.c: New routine to add annotation text to an image.
Requires a switch in CMakeLists.txt to link against Cairo library.

In deva-commandline.c (deva-filter, deva-visibility): Switch to basing
auto-clipping on multiple of median luminance, which is more robust than
the previous multiple of average luminance.  Added a compile-time option
for basing margin on only smaller dimension of input image.

In deva-commandline.c (deva-visibility): Added interface for more
measurement types.

In deva-canny.c:  Switch to basing luminance boundaries on log
luminance, not luminance itself.  This better approximates human
perception of contrast boundaries.  Switched to basing auto-thresholding
on local maxima of directional gradient, not all directional gradients.

In visualize-hazards.c/visualize-hazards.h: Cleaned up calling sequence.
Added in more measurement types.
 
In read-geometry.c, fixed possibly uninitialized DEVA_null_view.

Additional commenting.

version 3.0.01

Modified radiance-header.c to deal with indented VIEW records produced
by pcomb.  (pcomb has been modified in development version of Radiance
to output one non-indented VIEW record, alleviating the need for this
change if a recent enough version of Radiance is used to generate the
file.)

Cleaned up deva-gblur.c.  There should be no change in functionality.

Better error reporting in read-geometry routines.

Removed one extraneous routine from deva-margin.c

More extensive comments throughout.

Minor updates to man pages.

version 3.0.00

Added deva-visibility.  This extends the functionality of deva-filter to
provide estimates of likely low-vision hazards, defined as geometric
structures that may be a mobility hazard but are not co-located with
detectable visual contrast.  deva-visibility shares much of the
deva-filter code base.

Added deva-visualize-geometry, a utility program to display geometry
files used by deva-visibility.

Added make-coordinates-file, a support routine used to create one of the
input files needed by deva-visibility.

Added a build for Windows (see INSTALLATION file).  This requires a

make-coordinates-file.c
/*
 * Make a coordinates file specifing units of distance, coordinate system
 * orientation, and viewpoint for a set of DEVA geometry files.
 *
 * Units can be meters, centimeters, feet, or inches.  If desired, the
 * standard_units function can be used to converts these to a canonical
 * form (centimeters).  Note that the units apply to xyz and dst files,
 * but *not* to nor files with specify unit vectors.
 *
 * The coordinates file consists of two lines of text.  The first line
 * specifies the units.  The second line is a text representation of the
 * VIEW record taken from the cooresponding Radiance HDR file.
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include "radiance-header.h"
#include "read-geometry.h"
#include "deva-license.h"	/* DEVA open source license */

char	*Usage =
	    "make-coordinates-file units radiance-file.hdr coordinates-file";
int	args_needed = 3;

int
main ( int argc, char *argv[] )
{
    char		*unit;
    char		*radiance_filename;
    FILE		*radiance_fp;
    char		*coordinates_filename;
    FILE		*coordinates_fp;
    VIEW		view;
    int			argpt = 1;

    if ( ( argc - argpt ) != args_needed ) {
	fprintf ( stderr, "%s\n", Usage );
	return ( EXIT_FAILURE );	/* error exit */
    }

    unit = argv[argpt++];

    radiance_filename = argv[argpt++];

    radiance_fp = fopen ( radiance_filename, "r" );
    if ( radiance_fp == NULL ) {
	perror ( radiance_filename );
	exit ( EXIT_FAILURE );
    }

    coordinates_filename = argv[argpt++];

    coordinates_fp = fopen ( coordinates_filename, "w" );
    if ( coordinates_fp == NULL ) {
	perror ( coordinates_filename );
	exit ( EXIT_FAILURE );
    }

    if ( ( strcmp ( unit, "centimeters" ) != 0 ) &&
	    ( strcmp ( unit, "meters" ) != 0 ) &&
	    ( strcmp ( unit, "inches" ) != 0 ) &&
	    ( strcmp ( unit, "feet" ) != 0 ) ) {
	fprintf ( stderr, "invalid unit (%s)!\n", unit );
	exit ( EXIT_FAILURE );
    }

    /* get VIEW record from Radiance file header */
    DEVA_read_radiance_header ( radiance_fp, NULL, NULL, NULL, &view,
	    NULL, NULL, NULL );

    if ( view.type == 0 ) {
	fprintf ( stderr, "invalid or missing VIEW record!\n" );
	exit ( EXIT_FAILURE );
    }

    /* write two line coordinates file */
    fprintf ( coordinates_fp, "distance-units=%s\n", unit );

    fputs ( VIEWSTR, coordinates_fp );
    fprintview ( &view, coordinates_fp );
    putc ( '\n', coordinates_fp );

    return ( EXIT_SUCCESS );
}
